types/tx/service.go

```diff
@@ -0,0 +1,115 @@
+package tx
+
+import (
+	"context"
+	"encoding/hex"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/cosmos/cosmos-sdk/client"
+	codectypes "github.com/cosmos/cosmos-sdk/codec/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+)
+
+// BaseAppSimulateFn is the signature of the Baseapp#Simulate function.
+type BaseAppSimulateFn func(txBytes []byte) (sdk.GasInfo, *sdk.Result, error)
+
+type txServer struct {
+	clientCtx         client.Context
+	simulate          BaseAppSimulateFn
+	interfaceRegistry codectypes.InterfaceRegistry
+}
+
+// NewTxServer creates a new TxService server.
+func NewTxServer(clientCtx client.Context, simulate BaseAppSimulateFn, interfaceRegistry codectypes.InterfaceRegistry) ServiceServer {
+	return txServer{
+		clientCtx:         clientCtx,
+		simulate:          simulate,
+		interfaceRegistry: interfaceRegistry,
+	}
+}
+
+var _ ServiceServer = txServer{}
+
+// Simulate implements the ServiceServer.Simulate RPC method.
+func (s txServer) Simulate(ctx context.Context, req *SimulateRequest) (*SimulateResponse, error) {
+	if req.Tx == nil {
+		return nil, status.Error(codes.InvalidArgument, "invalid empty tx")
+	}
+
+	err := req.Tx.UnpackInterfaces(s.interfaceRegistry)
+	if err != nil {
+		return nil, err
+	}
+	txBytes, err := req.Tx.Marshal()
+	if err != nil {
+		return nil, err
+	}
+
+	gasInfo, result, err := s.simulate(txBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	return &SimulateResponse{
+		GasInfo: &gasInfo,
+		Result:  result,
+	}, nil
+}
+
+// GetTx implements the ServiceServer.GetTx RPC method.
+func (s txServer) GetTx(ctx context.Context, req *GetTxRequest) (*GetTxResponse, error) {
+	// We get hash as a hex string in the request, convert it to bytes.
+	hash, err := hex.DecodeString(req.Hash)
```

Comment:
are you sure it's not a hex representation of `HexBytes` because I think the format is different cc: @marbar3778

Reply:
I'm not 100% sure, would like to get confirmation from someone who has context.

(just a small note, tests seem to pass.)

Reply:
As per https://github.com/cosmos/cosmos-sdk/pull/7688#discussion_r514148966, we're going with hex.

I am still not sure whether this hex decoding needs to be changed or not though.